### ISSUE TITLE
Pointcut method designator does not allow to intercept private methods

### ISSUE THREAD

**Issue by lsascha** [neos/flow-development-collection#1982]
### Description

Pointcut method designator does not allow to intercept private methods,
only protected or public methods can be intercepted.

This is actually somewhat documented at
https://flowframework.readthedocs.io/en/stable/TheDefinitiveGuide/PartIII/AspectOrientedProgramming.html#method

since it only mentions public and protected methods but not private methods.

### Steps to Reproduce

1. create a class with a private method that is called.
2. Create an Aspect with an annotation like `@Flow\AfterReturning("method(Foo\Bar\Service\MyService->getResponse())")` pointing to the private method.

#### Expected behavior

The Aspect to be called when the private method returns something.

#### Actual behavior

Nothing happens.

**Comment by bwaidelich** [comment]
For this to work we would have to change the visibility of the method in the proxy I guess:
```php
class SomeClass {
  private function someMethod() {
  }
}
```
would become
```php
class SomeClass_Original {
  protected function someMethod() {
  }
}

class SomeClass extends SomeClass implements ProxyInterface {
  private function someMethod() {
  }
}
```

We do that already for `final` classes / methods and turn:
```php
class SomeClass {
  final public function someMethod() {
  }
}
```
into
```php
class SomeClass_Original {
  /*final*/ public function someMethod() {
  }
}

class SomeClass extends SomeClass implements ProxyInterface {
  final public function someMethod() {
  }
}
```

see https://github.com/neos/flow-development-collection/blob/master/Neos.Flow/Classes/ObjectManagement/Proxy/Compiler.php#L245-L253

We could do the same with the visibility modifier. But I'm not sure about the implications since the semantics change if you change a method from private to protected..


I'd rather go for an exception during compilation in that case

**Comment by albe** [comment]
Yep, that was my plan also. Though the idea to level up the visibility could also work - which would be better since there are no exceptions to AOP any more then which need to be documented. But if at all that would be a FEATURE - for now, let's just throw an exception as a BUGFIX to improve the DX.

**Comment by bwaidelich** [comment]
> the idea to level up the visibility could also work

What do you think about the breakiness of this?
i.e. private methods behave differently when serializing.. Seems like a dangerous magic

**Comment by albe** [comment]
`private methods behave differently when serializing` - uh, I'm not even aware of that 😨 Under what circumstance would that take effect for a proxied class?

My thinking was/is, that the method is still private inside the proxy class - so in theory from the outside nothing should change.

**Comment by bwaidelich** [comment]
Ah, sorry I mixed things up.. Serializing affects only private _fields_ of course – those behave differently.

> My thinking was/is, that the method is still private inside the proxy class

That was a typo in my example at first. You can only ever _relax_ the visibility in a subclass (i.e. `private` => `protected` works, but not the other way around).

So, other then the `final` keyword that we can keep in the proxy class the private method would become protected and I'm not sure if that has other implications..

And thinking about it.. Is it really sensible to advice a private method? It's very likely to change

**Comment by kitsunet** [comment]
Note, we do not touch the original class code (apart from the class name) so far, so that would be a pretty big change to the proxy creation IMHO.

**Comment by bwaidelich** [comment]
> Note, we do not touch the original class code (apart from the class name)

Yes, we do for the `final` keyword, see comments above. But I agree that this is a different case

**Comment by kitsunet** [comment]
right, grrrr ;)

**Comment by albe** [comment]
> You can only ever relax the visibility in a subclass

You are correct :)

> for now, let's just throw an exception as a BUGFIX to improve the DX.

Repeating myself just to get the point across :)

**Comment by albe** [comment]
> Is it really sensible to advice a private method? It's very likely to change

I mean, you could say the same of protected methods and injecting code around those from the outside is already bending limits. So it's pretty arbitrary of us to say "private may not be adviced" (apart from the technical problems of getting it done) - I mean, advicing things should always be a last resort and there are probably $reasons - e.g. logging invocations of internal methods.

So given that, I'd gladly see someone enhance the AOP proxy builder to juggle the visibility around to make this work.

**Comment by bwaidelich** [comment]
> I mean, you could say the same of protected methods

What I meant is:

```php
class SomeClass
{
   private function thisIsPrivate() { ... }

   protected function thisIsProtected() { ... }
}
```

there's a fundamental difference between the two methods since the first one is truly private to this class while the other one is shared with all sub classes – similarly with private/protected properties.
Or, to put it differently: changing a public or protected field is a potentially breaking change while changing a private one should not.

And if we have the choice between:
* We change the visibility from `private` to `protected` via proxy building
* We disallow to advice non-existent or private fields

(there might be more options!?)

I'd go for the latter because the former might have unplanned side effects, for example the different serialization behavior:

```php
class SomeClass {
    protected $someProperty = 'foo';
}
echo serialize(new SomeClass());
```

=>
```
O:9:"SomeClass":1:{s:15:"*someProperty";s:3:"foo";}
```

and with private:

```
O:9:"SomeClass":1:{s:23:"SomeClasssomeProperty";s:3:"foo";}
```

> apart from the technical problems of getting it done

It's actually quite easy I just found out (see #2036 - WIP because it misses doc comments and a test, anyone feels like adding those?)

**Comment by albe** [comment]
Guess you overlooked #2035 :D But interresting to see that you would throw the exception at a completely different place.

> I'd go for the latter because the former might have unplanned side effects, for example the different serialization behavior:
> ```php
> class SomeClass {
>     protected $someProperty = 'foo';
> }
> echo serialize(new SomeClass());
> ```

Well, do we actually advise properties? We introduce new ones, but otherwise? So this would only affect methods which are not serialized. But even if, the proxied class would rather look like this:

```php
class SomeClass_Original {
    protected function foo() { ... }
}

class SomeClass extends SomeClass_Original {
    private function foo() { ... }
}
```
and that private `foo` would internally invoke `SomeClass_Original::foo` via `call_user_func_array`. So, unless someone directly interacts with the `_Original` class, the change should not be observable from the outside, unless I'm completely missing something.

> It's actually quite easy I just found out

So we would just need to set `overrideMethodVisibility('protected')` in the proxy builder? But we'd need to change the visibility of the original class, not the proxy class.

**Comment by bwaidelich** [comment]
> Guess you overlooked #2035

Doh, indeed – and I think I even looked at it before.. But IMO it would make sense to enforce that during compliation time already

> do we actually advise properties?

The serialization example was just a case of private/protected behaving differently in some cases.

> [...] the change should not be observable from the outside

OK, if that's really the case I obviously don't have a major issue with enabling AOP for private methods (although doing so seems like dangerous practice to me)

>> It's actually quite easy I just found out

I meant it was easy to _disallow_ advising of private methods, I think I got you wrong there.

> So we would just need to set overrideMethodVisibility('protected') in the proxy builder?

I think we have to do something similar to the replaced `final` keyword in proxied classes: https://github.com/neos/flow-development-collection/blob/53c9a7d3d877fc0eb5b2f08d636910212c42b74c/Neos.Flow/Classes/ObjectManagement/Proxy/Compiler.php#L247-L253

But I'm afraid this would require a bigger rewrite since we don't want to make every private method protected IMO but only the adviced ones.. Would be a chance to skip proxy building for classes without advices... :)

**Comment by albe** [comment]
> But I'm afraid this would require a bigger rewrite

Okay, so we're back at "technical problems of getting it done" :)

> > Guess you overlooked #2035
> Doh, indeed – and I think I even looked at it before.. But IMO it would make sense to enforce that during compliation time already

Hm. I might forget something, but I was under the impression the ProxyBuilder (so compile time) used the PointcutFilter to find all methods that need to be proxied.

**Comment by bwaidelich** [comment]
Hey guys,
how should we proceed with this one?

I would consider #2036 (or #2035) an improvement to the current situation. We could always still add _support_ for proxied private methods if we figure out a good way to implement this. Personally, I won't take care of that because I still think that it should(tm) not be required. But obviously I'd be happy if someone else took care.
Otherwise I'd go ahead an add documentation and tests to my PR

**Comment by bwaidelich** [comment]
@albe I finally came around to take a closer look on to your implementation (#2035) but I couldn't get it to work.

So I recreated my fix (#2131) with the `5.3` target branch, tweaked it a little, copied some of your exception message and added doc comments..
